### PR TITLE
Add support to opt-in for using ENI's primary IP for allocations

### PR DESCRIPTION
**Pull request by hemanthmalla**
Recently ENI's primary IP has been removed from pool of IPs available
for allocations on the node. This has been done to make sure SNAT works
correctly when kube-proxy is used. See https://github.com/cilium/cilium/pull/15453/commits/bc0a748021f5234880fdede51cf04e4fe6d1833b
for more details. However this can lead to IP wastage in large clusters.
The SNAT case does not apply when cilium runs in kube proxy replacement
mode.

This commit adds a new flag called `--aws-use-primary-address` to the
operator and new field in ciliumnode CRD ENI spec called
UseENIPrimaryAddress to opt in for using the ENI primary address. Azure
IPAM mode already has a flag like this.

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

### CODE REVIEW

File: pkg/nodediscovery/nodediscovery.go
Comment (gandro) [inline pkg/nodediscovery/nodediscovery.go]: Why this removal? Looks like an accidental change

File: operator/option/config.go
Comment (gandro) [inline operator/option/config.go:526]: Is this flag still necessary now that you populate the CRD field via CNI conf?

File: pkg/ipam/crd_eni.go
Comment (gandro) [inline pkg/ipam/crd_eni.go]: nit: The overall logic of the agent side of things looks sound. But you could add the `useENIPrimaryIP` boolean to `eniDeviceConfig` instead. Or maybe just check if `cfg.ip` is not nil when setting the `cfp.ip` on the link.

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks! Looks good overall, but I left a few comments

File: pkg/nodediscovery/nodediscovery.go
Comment (hemanthmalla) [reply pkg/nodediscovery/nodediscovery.go]: Its is an accidental change from cherry picking changes between 1.10 and master. Sorry about this. Will push a fix shortly.

File: pkg/ipam/crd_eni.go
Comment (hemanthmalla) [reply pkg/ipam/crd_eni.go]: Good idea. That sounds cleaner. This change would need `parseENIConfig`'s function signature to be updated though. Will make this change.

File: operator/option/config.go
Comment (hemanthmalla) [reply operator/option/config.go:526]: So, I could not find a clean way to propagate the flag from the CRD spec in places like 

https://github.com/cilium/cilium/blob/d23d3d366d72124537d8d0ff3c8df5acd494f854/pkg/aws/ec2/ec2.go#L327-L330

Do you have ideas on how this can be done ? This is tricky because `GetInstances()` is listing interfaces directly and does not access the node object.

**Review by christarazi — APPROVED** [review]
LGTM pending Sebastian's comments and fixing up the Travis failure for regenerating `helm-values` file.

File: operator/option/config.go
Comment (gandro) [reply operator/option/config.go:526]: Ah, I see. Yeah, in that case a global flag is probably the better option for the operator, since we can't really set the `usePrimary` flag per node (as we can with the other CNI-populated CRD fields).

**Review by gandro — APPROVED** [review]
Awesome, thanks!

**Review by twpayne — APPROVED** [review]

**Comment by hemanthmalla** [comment]
Marking as ready to merge since we have reviews from ipam and kubernetes. GKE failures are not relevant.

**Pull request merged**
